stomp_moveit/src/stomp_smoothing_adapter.cpp

```diff
@@ -0,0 +1,139 @@
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Raghavender Sahdev nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Henning Kayser */
+
+// ROS
+#include <ros/ros.h>
+#include <class_loader/class_loader.hpp>
+
+// MoveIt
+#include <moveit/planning_request_adapter/planning_request_adapter.h>
+#include <moveit/planning_interface/planning_interface.h>
+
+// STOMP
+#include <stomp_moveit/stomp_planner.h>
+
+namespace stomp_moveit
+{
+class StompSmoothingAdapter : public planning_request_adapter::PlanningRequestAdapter
+{
+public:
+  StompSmoothingAdapter() : planning_request_adapter::PlanningRequestAdapter()
+  {
+  }
+
+  virtual void initialize(const ros::NodeHandle& node_handle) override
+  {
+    ros::NodeHandle nh(node_handle);
+    if (!StompPlanner::getConfigData(nh, group_config_))
+      ROS_ERROR("Unable to find valid group config for StompSmoothingAdapter");
+  }
+
+  virtual std::string getDescription() const override
+  {
+    return "Stomp Smoothing Adapter";
+  }
+
+  bool adaptAndPlan(const PlannerFn& planner, const planning_scene::PlanningSceneConstPtr& ps,
+                    const planning_interface::MotionPlanRequest& req, planning_interface::MotionPlanResponse& res,
+                    std::vector<std::size_t>& added_path_index) const override
+  {
+    // Following call to planner() calls the OMPL planner and stores the trajectory inside the MotionPlanResponse res
+    // variable which is then used by STOMP for optimization of the computed trajectory
+    if (!planner(ps, req, res))
```

Comment:
Does this call the planner directly or all the preceding adapters followed by the planner?

Reply:
Planning adapters are called in order of the adapter chain specified for a planner. `PlannerFn` is a generic wrapper that either calls the main planner like OMPL or any other planning adapter in the chain. http://docs.ros.org/indigo/api/moveit_core/html/planning__request__adapter_8cpp_source.html.

Reply:
Thanks for clarifying, that's what I thought.